# Machine Learning Basics with Scikit-learn: Day 1

## Introduction

### Objectives
The workshop will focus on the basics of *scikit-learn*, which is one of the most popular machine learning libraries in Python. After the workshop you will:

* Understand how to use scikit-learn functions and documentation.
* Learn the usual procedure to transform your data for a machine learning model.
* Create basic machine learning models.

### Why scikit-learn?
Scikit-learn is an open source machine learning library that supports supervised and unsupervised learning. It also provides various tools for model fitting, data preprocessing, model selection and evaluation, and many other utilities. It has a standardized and simple interface for preprocessing data and model training, optimization and evaluation. 

### Why are we teaching machine learning with scikit-learn?
Using machine learning is becoming a mandatory new skill for many professional workers. As work and organizations demand more use of data analyses, machine learning allows us to analyze large datasets to extract meaningful information. Although machine learning techniques have been there for decades, modern programming languages are making these techniques available for thousands of users. Rather than developing each method from scratch, *scikit-learn* offers a simple way to implement them with our datasets. This package provides a easy toolkit to master and leverage machine learning skills. We hope that learning sckikit-learn helps you comprehend the overall process of data transformation: from curating and importing datasets, to curating models for data knowledge purposes. 

### Structure of the Workshop
The workshop is divided into 5 days:

1. Introduction to scikit-learn
2. Supervisded learning: Classification models
3. Unsupervised learning: Clustering models
4. Data cleaning / transformation
5. Model selection

Today, we will start with an overview of *scikit-learn.* We will load datasets, create training and testing datasets, create a model, and evaluate it. We will delve into the functions, models, and details these days. However, keep in mind that this workshop's content will stay at an introductory level. If you are interested in learning more, here are some good resources that you can explore:
* [Official documentation](https://scikit-learn.org/stable/index.html)
* [Google Cloud AI Adventures](https://www.youtube.com/hashtag/aiadventures)
* [Google's Machine Learning Crash Course](https://developers.google.com/machine-learning/crash-course)
* [Stanford University: CS229: Machine Learning](http://cs229.stanford.edu/)

## Installing scikit-learn

Conda and Google Colab already have scikit-learn. You should skip the instalation if you're using one of those. If you are running this notebook on your own environment, please run the following command. And for more instructions, please check out the [documentation](https://scikit-learn.org/stable/install.html)

In [1]:
# If you do not have scikit-learn installed, uncomment the following lines and run the following command. 
# import sys
# !{sys.executable} -m pip install scikit-learn==0.24.2
# !{sys.executable} -m pip install scikit-learn --upgrade

Let's check that the package is in your environment. Run the following command.

In [2]:
import sklearn
sklearn.show_versions()


System:
    python: 3.9.1 (default, Feb  1 2021, 20:41:56)  [Clang 12.0.0 (clang-1200.0.32.29)]
executable: /usr/local/opt/python@3.9/bin/python3.9
   machine: macOS-10.15.7-x86_64-i386-64bit

Python dependencies:
          pip: 21.1.3
   setuptools: 52.0.0
      sklearn: 0.24.1
        numpy: 1.20.0
        scipy: 1.6.1
       Cython: None
       pandas: 1.2.1
   matplotlib: 3.3.4
       joblib: 1.0.1
threadpoolctl: 2.1.0

Built with OpenMP: True


## The big picture
One of the biggest advantages of machine learning is to determine how to differentiate observations using a computational model, rather than using human coding and manual rules. When we have thousands of obsevations, machine learning models helps us automatize, scale, and guarantee the reproducibility these coding processes. The basic steps are:
1. Gathering data
2. Preparing that data
3. Choosing a model
4. Training
5. Evaluation
6. Prediction.

## 1. Gathering data

The first step is collecting data and understanding the dataset. This step is very important because the quality and quantity of data that you gather will directly determine how good your predictive model can be. Also, many machine learning models are *biased* because of the data. A well-known example is [this face-generator model](https://www.theverge.com/21298762/face-depixelizer-ai-machine-learning-tool-pulse-stylegan-obama-bias) that did not recognize President Obama as an African-American person. 

You need to collect data and take into account these recommendations:
* **Number of observations (*N*):** In statistics, there is a theorem called "Law of Big Numbers." According to this law, the average of the results obtained from a large number of trials should be close to the expected value. In other words, by having a large number of observations, your model will tend to become closer to the expected value. Creating machine learning models is based on *big datasets.* Imagine companies that collect big data from their clients and are able to get a clear picture of them. Prediciton becomes more accurate as long as you have more observations and knowledge of your population. 
* **Missing data**: Many observations may lack some values. You can have multiple reasons for this: data-collection issues, restricted data, information not available, etc. It is important to make strategies whenever you have missing data. 
* **Representativity**: One main problem in most datasets is checking how representative is the dataset with respect to the population. How can we be sure that the dataset is not baised? Computers do not understand the dataset context, or how the dataset was collected. This problem lies on the people who collected the data. Moreover, people who got the data must be aware about any potential flaws or inequealities in the dataset. Descriptive analysis should guide any checkings and validation processes. 

In this workshop, we will use the datasets provided by *scikit-learn*. You can check other [toy datasets](https://scikit-learn.org/stable/datasets/toy_dataset.html) if you are interested in exploring more. These datasets are well-known, public, and used frequently for learning and testing purposes.

### Loading datasets

Before we get started, some terms that we must get familiar with:
* **Samples**: A sample is an observation available in the dataset. Also, they are known as observations, records, etc. They are usually the rows of a dataset table.
* **Features**: A feature is an individual measurable property. We also know them as variables and attributes. Usually, these are the columns of a dataset table.

We will start importing the scikit-learn's datasets:

In [3]:
from sklearn import datasets

All datasets are now in the environment. We can call specifically one of those and assign them as a variable. We will use the *Boston house prices dataset.* Each record in the database describes a Boston suburb or town. The data was drawn from the Boston Standard Metropolitan Statistical Area (SMSA) in 1970. The attributes are defined as follows (taken from the UCI Machine Learning Repository1): CRIM: per capita crime rate by town

* CRIM: crime per capita crime rate by town
* ZN: proportion of residential land zoned for lots over 25,000 sq.ft.
* INDUS: proportion of non-retail business acres per town
* CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
* NOX: nitric oxides concentration (parts per 10 million)
* RM: average number of rooms per dwelling
* AGE: proportion of owner-occupied units built prior to 1940
* DIS: weighted distances to five Boston employment centres
* RAD: index of accessibility to radial highways
* TAX: full-value property-tax rate per \$10,000
* PTRATIO: pupil-teacher ratio by town
* B: 1000(Bk - 0.63)^2 where Bk is the proportion of black people by town
* LSTAT: proportion of lower status of the population
* MEDV: Median value of owner-occupied homes in $1000’s

This model has been frequently used for regression models.

In [4]:
boston = datasets.load_boston()

## 2. Preparing the data

In a machine learning model, we will have *features* and a *target*. The target is the variable to predict/estimate by the model. The value of the target depends on the features' values. Usually, targets are the outcomes of a process (e.g., giving a loan, earnings). In statistics, the target is also known as the *dependent variable*, and the features are the *independent variables*.

![In this example, demographic informations are used to predict users' behavior when they are navigating on a website](https://d2m6ke2px6quvq.cloudfront.net/uploads/2020/09/11/0e1df989-5fc9-474b-ba49-5eaebfc2d795.png)

In this example, users' demographic and activity information are used to predict users' behavior when they are navigating on a website.

### Getting the features and target data from scikit-learn
*scikit-learn* datasets are formated as dictionary-like objects. They hold all the data and some metadata about the data. These data are stored in the `.data` member, which is a `n_samples`, `n_features` array. 

The member `.feature_names` will contain the names recorded for each feature.

In the case of supervised problem, one or more response variables are stored in the `.target` member. These are variables to *predict*.

Let's check the Boston dataset:

In [5]:
print(boston.data)

[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In this dataset, the **Median Value** (the last attribute) is usually the target.

In [6]:
print(boston.target)

[24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.  18.9 21.7 20.4
 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8
 18.4 21.  12.7 14.5 13.2 13.1 13.5 18.9 20.  21.  24.7 30.8 34.9 26.6
 25.3 24.7 21.2 19.3 20.  16.6 14.4 19.4 19.7 20.5 25.  23.4 18.9 35.4
 24.7 31.6 23.3 19.6 18.7 16.  22.2 25.  33.  23.5 19.4 22.  17.4 20.9
 24.2 21.7 22.8 23.4 24.1 21.4 20.  20.8 21.2 20.3 28.  23.9 24.8 22.9
 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.  22.9 25.  20.6 28.4 21.4 38.7
 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8
 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.  20.3 20.5 17.3 18.8 21.4
 15.7 16.2 18.  14.3 19.2 19.6 23.  18.4 15.6 18.1 17.4 17.1 13.3 17.8
 14.  14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4
 17.  15.6 13.1 41.3 24.3 23.3 27.  50.  50.  50.  22.7 25.  50.  23.8
 23.8 22.3 17.4 19.1 23.1 23.6 22.6 29.4 23.2 24.6 29.9 37.2 39.8 36.2
 37.9 32.5 26.4 29.6 50.  32.  29.8 34.9 37.  30.5 36.4 31.1 29.1 50.
 33.3 3

In [7]:
print(boston.feature_names)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


Before applying any model, it is a good idea to convert this data structure to a pandas dataframe. We can keeep the data in one single object and keep the features' names. We can do by calling the dataframe on `boston.data` and adding the target variable to the dataframe from `boston.target`

In [8]:
import pandas as pd
boston_df = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_df['PRICE'] = boston.target

We will print the first 5 columns of this pandas dataframe. We will see each column with their respective feature name.

In [9]:
print(boston_df.head())

      CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  PRICE  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2  


### Exercise 1
Before we create a model, let's get familiar with the target column, "PRICE," which is the value of prices of the houses. Run the following command and check the mean, minimum value, maximum value, and the 50\%. How is the data distributed?

In [10]:
print(boston_df['PRICE'].describe())

count    506.000000
mean      22.532806
std        9.197104
min        5.000000
25%       17.025000
50%       21.200000
75%       25.000000
max       50.000000
Name: PRICE, dtype: float64


## 3. Training and Testing sets

To see if the model is capable to predict "future" values, we split the original dataset in two sets: one portion will be used for **training**, and the second portion will be for **testing** purposes. The testing dataset will *test* the model against data that has never been used for *training*. Having a testing dataset allows us to see how the model might perform against data that it has not yet seen. This is meant to be representative of how the model might perform in the real world.

A rule of thumb is use for a training-evaluation split somewhere on the order of 80/20, 90/10. Much of this depends on the size of the original source dataset. We will start training the model with 80% of the sample and test the model with the remaining 20%. We do this to assess the model's performance on unseen data.

We must separate the features that will act as independent variables (`X`) from the target (`Y`). The independent variables include all attributes but `'PRICE'`:

In [11]:
X = boston_df.drop('PRICE', axis = 1)
y = boston_df['PRICE']

To split the data, we use `train_test_split` function provided by *scikit-learn* library. We finally print the shapes of our training and test set to verify if the splitting has occurred properly.

In [12]:
# Import the train_test_split function
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

Now, we check the dimensions of both sets are correct. Since we have 506 observations, the training set has ~404 observations (80%) and the testing set has about 102 observations. Since we have 13 features in total, each `X` dataframe has 13 columns:

In [14]:
print(X_train.shape)
print(X_test.shape)

(404, 13)
(102, 13)


Finally, the target is a single column with 404 observations for the training set, and 102 observations for the testing set.

In [15]:
print(y_train.shape)
print(y_test.shape)

(404,)
(102,)


### Exercise 2
Change the proportion for the training set. Instead of 80\%, set it up for **90%** (i.e., 90% for training and 10% for testing). How many observations would you have for the traning and testing datasets?

In [16]:
## Run the code here
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(X, y, test_size = 0.1)

In [17]:
# Print X datasets' shape
print(X_train_2.shape)
print(X_test_2.shape)

(455, 13)
(51, 13)


In [18]:
# Print y datasets' shape
print(y_train_2.shape)
print(y_test_2.shape)

(455,)
(51,)


## 4. Choosing a model

The next step in our workflow is choosing a model. There are many models that researchers and data scientists have created over the years. Some are very well suited for image data, others for sequences (like text, or music), some for numerical data, others for text-based data. 

Since we have 13 features, one target, and numerical data, we can use a **linear regression model.** Regression models are very useful for numeric datasets.

In [19]:
# Import Linear Regression model
from sklearn.linear_model import LinearRegression

In [20]:
reg = LinearRegression()
reg.fit(X_train, y_train)

LinearRegression()

### Exercise 3
Run again the lineal regression model using your training set with 90% of the observations (`X_train_2`,`y_train_2`). Call your model `reg2`.

In [21]:
# Run regression model
reg2 = LinearRegression()
reg2.fit(X_train_2, y_train_2) 

LinearRegression()

## 5. Evaluating your model

Once training is complete, it's time to see if the model provides accurate results. A good practice is to compare the original values (`y_train`) with the predicted values given by the model. We want to check how *good* is the model predicitng the values were used for training. 

We use the function `predict` to get the values according to the trained model.

In [22]:
y_train_predict = reg.predict(X_train)

We compare the predicted training values with the real training values by checking their difference. If the difference is small, the model is predicting values close to the real ones. We will create a table now to compare them visually.

In [23]:
y_train_df = pd.DataFrame({'y_real': y_train, 'y_predicted': y_train_predict})
y_train_df['difference'] = y_train_df.y_real - y_train_df.y_predicted 
y_train_df.head(10)

y_real  y_predicted  difference
16     23.1    20.842998    2.257002
368    50.0    22.971757   27.028243
114    18.5    25.020635   -6.520635
357    21.7    21.559434    0.140566
213    28.1    25.634705    2.465295
106    19.5    17.145487    2.354513
286    20.1    20.072545    0.027455
174    22.6    26.493185   -3.893185
396    12.5    19.112699   -6.612699
228    46.7    36.017114   10.682886

As expected, some predicted values are close to the real values (i.e., their differences are close to zero), and some others are far from expected (i.e., big values in their differences). 

We need **metrics** to evaluate empirically how good is this model. For linear regression models, one metric is the *coefficient of determination* (R^2) of the prediction. This score is related to the differences between the predicted values and the original values. The best possible score is 1.0. For more details, click [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). 

In [24]:
r2 = round(reg.score(X_train, y_train),2)
print('R2 score is {}'.format(r2))

R2 score is 0.74


### Exercise 4
Calcualte the predicted value using the model `reg2` with the 90% training dataset. Use `y_train_predict_2` for the new predicted values. Then, copy the pandas dataframe and replace the files with your 90% datasets (`y_train_2` and `y_train_predict_2`). The new dataframe should be called `y_train_df_2`. Print `y_train_df_2` dataframe's head. 

In [25]:
# Calculate the predicted prices based on the 90% training set
y_train_predict_2 = reg2.predict(X_train_2)

# Create the dataframe
y_train_df_2 = pd.DataFrame({'y_real': y_train_2, 'y_predicted': y_train_predict_2})
y_train_df_2['difference'] = y_train_df_2.y_real - y_train_df_2.y_predicted 
y_train_df_2.head(10)

y_real  y_predicted  difference
311    22.1    26.612465   -4.512465
60     18.7    18.045837    0.654163
128    18.0    18.672194   -0.672194
298    22.5    29.160456   -6.660456
391    23.2    16.952115    6.247885
51     20.5    24.280924   -3.780924
110    21.7    20.254179    1.445821
2      34.7    30.780529    3.919471
176    23.2    25.492126   -2.292126
350    22.9    20.521041    2.378959

Compute the R^2 of your model and compare it with the previous model' R^2. Is it better or worse?

In [26]:
r2 = round(reg.score(X_train_2, y_train_2),2)
print('R2 score is {}'.format(r2))

R2 score is 0.75


## 6. Prediction
In this final step, we use the model that we trained to *predict* new values (or values that were not tested before). This is the final test to check how good the model is. If the phenomena or casual effects are reflected correctly in our model, then the model will be capable to predict new observations. If the model's performance gets worse, we must reconsider the training dataset, features, and machine learning model used. 

Like the prior step, we first start predicting the target values (`y_test_predict`) from the testing dataset (`X_test`)

In [27]:
y_test_predict = reg.predict(X_test)

We compare the predicted testing values with the real testing values by checking their difference. If the difference is small, the model is predicting values close to the real ones. We will create a table now to compare the numbers.

In [28]:
y_test_df = pd.DataFrame({'y_original': y_test, 'y_predicted': y_test_predict})
y_test_df['difference'] = y_test_df.y_original - y_test_df.y_predicted 
y_test_df.head(10)

y_original  y_predicted  difference
349        26.6    22.063683    4.536317
502        20.6    22.441019   -1.841019
81         23.9    26.729543   -2.829543
372        50.0    24.694124   25.305876
472        23.2    22.243383    0.956617
9          18.9    18.662283    0.237717
487        20.6    21.217312   -0.617312
153        19.4    17.381433    2.018567
339        19.0    21.455078   -2.455078
86         22.5    22.433027    0.066973

Finally, we compute the R^2 of this model and check its performance. We can expect to see a lower performance compared to the training validation. We will discuss in another session how we can make these models have better scores. 

In [29]:
r2 = round(reg.score(X_test, y_test),2)
print("R^2: {}".format(r2))

R^2: 0.73


### Exercise 5

Calcualte the predicted value using the model `reg2` with the 90% testing dataset. Use `y_test_predict_2` for the new predicted values. Then, copy the pandas dataframe and replace the variables with your 90% testing dataset (`y_test_2` and `y_test_predict_2`). The new dataframe should be called `y_test_df_2`. Print `y_test_df_2` dataframe's head. 

In [30]:
# Calculate the predicted value
y_test_predict_2 = reg2.predict(X_test_2)

In [31]:
# Create the dataframe
y_test_df_2 = pd.DataFrame({'y_original': y_test_2, 'y_predicted': y_test_predict_2})
y_test_df_2['difference'] = y_test_df_2.y_original - y_test_df_2.y_predicted 
y_test_df_2.head(10)

y_original  y_predicted  difference
170        17.4    22.749264   -5.349264
11         18.9    21.955882   -3.055882
469        20.1    18.187989    1.912011
156        13.1    13.606285   -0.506285
304        36.1    33.140070    2.959930
266        30.7    31.379309   -0.679309
29         21.0    20.943495    0.056505
102        18.6    20.061146   -1.461146
165        25.0    25.650152   -0.650152
308        22.8    28.732440   -5.932440

Compute the R2 of your model and compare it with the previous model' R2. Is it better or worse?

In [32]:
r2 = round(reg2.score(X_test_2, y_test_2),2)
print('R2 score is {}'.format(r2))

R2 score is 0.68
